---
title: "Zeisel et al. scRNA-seq Example"
author: "Philippe Boileau"
output:
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
# libraries
library(here)
library(dplyr)
library(scRNAseq)
library(scater)
library(org.Mm.eg.db)
library(scran)
library(future)
library(future.apply)
library(cvCovEst)
library(RSpectra)
library(BiocSingular)
library(cluster)
library(ggpubr)


# helper functions
source(file = here("utils", "plot_utils.R"))

# knitr options
knitr::opts_chunk$set(echo = TRUE)
```


We apply `cvCovEst` to the Zeisel et al. single-cell transcriptomics dataset
contained in the `scRNAseq` Bioconductor package. The dataset preparation
closely follows that contained in Chapter 9 of
[*Orchestrating Single-Cell Analysis*](https://osca.bioconductor.org/dimensionality-reduction.html).

# Data Preparation

## Load Data

```{r load-data, message=FALSE, results='hide'}
# load the dataset
sce_zeisel <- ZeiselBrainData()

# aggregate some gene's expressions
sce_zeisel <- aggregateAcrossFeatures(
  sce_zeisel,
  sub("_loc[0-9]+$", "", rownames(sce_zeisel))
)

# annotate the genes
rowData(sce_zeisel)$Ensembl <- mapIds(
  org.Mm.eg.db, keys = rownames(sce_zeisel),
  keytype = "SYMBOL", column = "ENSEMBL"
)
```

## Quality Control

```{r qc}
# compute quality control metrics
stats <- perCellQCMetrics(
  sce_zeisel,
  subsets = list(
    Mt = rowData(sce_zeisel)$featureType == "mito"
  )
)
qc <- quickPerCellQC(
  stats,
  percent_subsets = c("altexps_ERCC_percent", "subsets_Mt_percent")
)

# discard poor quality cells
sce_zeisel <- sce_zeisel[, !qc$discard]
```

## Normalization

```{r normalization}
# normalize the data
set.seed(412321)
clusters <- quickCluster(sce_zeisel)
sce_zeisel <- computeSumFactors(sce_zeisel, clusters = clusters)
sce_zeisel <- logNormCounts(sce_zeisel)
```

## Variance Modelling

Identify top 1000 most variable genes.

```{r mod-var}
dec_zeisel <- modelGeneVarWithSpikes(sce_zeisel, "ERCC")
top_hvgs <- getTopHVGs(dec_zeisel, n = 1000)
```

```{r cleanup, echo=FALSE}
rm(dec_zeisel, qc, stats, clusters)
```


# Apply `cvCovEst`

We now apply the `cvCovEst` function to the matrix of log-normalized cells to
estimate its covariance matrix. Note that we center and scale the data as to
estimate the sample correlation matrix.

```{r cvCovEst-corr-mat-1000, eval=FALSE}
# format data
data_in <- t(logcounts(sce_zeisel)[which(rownames(sce_zeisel) %in% top_hvgs), ])

# run cvCovEst
set.seed(5234523)
plan(multiprocess, workers = 5L)
cv_cov_results <- cvCovEst(
  dat = data_in,
  estimators = c(linearShrinkLWEst, thresholdingEst, sampleCovEst,
                 scadEst, poetEst, adaptiveLassoEst, nlShrinkLWEst),
  estimator_params = list(
    thresholdingEst = list(gamma = seq(0.05, 0.3, by = 0.05)),
    scadEst = list(lambda = seq(0.05, 0.5, by = 0.05)),
    poetEst = list(k = seq(5L, 10L), lambda = seq(0.05, 0.3, by = 0.05)),
    adaptiveLassoEst = list(lambda = seq(0.1, 0.5, by = 0.1),
                            n = seq(0.1, 0.5, by = 0.1))
  ),
  cv_loss = cvMatrixFrobeniusLoss,
  cv_scheme = "v_fold",
  v_folds = 5L,
  center = TRUE,
  scale = TRUE,
  parallel = TRUE
)

# save the results
saveRDS(cv_cov_results, file = here("zeisel-example",
                                    "results", "p1000_cv_cov_est_results.rds"))
```

```{r cvCovEst-pc-1000a}
cv_cov_results <- readRDS(
  file = here("zeisel_example", "results", "p1000_cv_cov_est_results.rds")
)

# now perform PCA with cvCovEst estimate
eig_decomp <- eigs_sym(cv_cov_results$estimate, k = 20)

data_in <- t(logcounts(sce_zeisel)[which(rownames(sce_zeisel)
                                         %in% top_hvgs), ])
cv_cov_est_pca <- data_in %*% eig_decomp$vectors
reducedDim(sce_zeisel, "cvPCA") <- cv_cov_est_pca
```


### Apply PCA

```{r run-pca-1000}
sce_zeisel <- runPCA(sce_zeisel, subset_row = top_hvgs, scale = TRUE)
reducedDim(sce_zeisel, "PCA") <- reducedDim(sce_zeisel, "PCA")[, 1:20]
```

### Compare UMAP

```{r umap-plots-1000, fig.width=10, fig.height=4}
set.seed(12312)

# regular PCA as initialization
sce_zeisel <- runUMAP(sce_zeisel, dimred = "PCA")

# cvPCA as initialization
sce_zeisel <- runUMAP(sce_zeisel, dimred = "cvPCA", name = "cvUMAP")

# plot the comparison
combo_p <- plot_reduced_dims(sce_zeisel, "level1class")
combo_p

saveRDS(
  combo_p,
  file = here("combined", "graphs", "zeisel_1000hvgs.rds")
)
```
